File: swift/15-3Sum.swift
class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        let sortedArray = nums.sorted()
        var result: [[Int]] = []
        for i in 0..<sortedArray.count {
            if i > 0 && sortedArray[i] == sortedArray[i - 1] { continue }
            var l = i + 1
            var r = sortedArray.count - 1
            while l < r {
                var threeSum = sortedArray[i] + sortedArray[l] + sortedArray[r]
                if threeSum > 0 {
                    r -= 1
                } else if threeSum < 0 {
                    l += 1
                } else {
                    result.append([sortedArray[i], sortedArray[l], sortedArray[r]])
                    l += 1
                    while sortedArray[l] == sortedArray[l - 1] && l < r {
                        l += 1
                    }
                }
            }
        }
        return result
    }
}

File: kotlin/15-3Sum.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        if(nums.size<3){
            return List(0){List<Int>(0){0}}
        }
        if(nums.size==3){
            return if(nums.sum()==0) listOf(nums.toList()) else List(0){List<Int>(0){0}}
        }
        nums.sort()
        val a = MutableList(0){List(3){0}}
        val s = nums.size
        var j = 0
        var k = 0
        
        for(i in 0 until s-2){
            if(i!=0 && nums[i]==nums[i-1]){
                continue
            }
            j = i+1
            k = s-1
            while(j<k){
                if(nums[i]+nums[j]+nums[k]>0){
                    k -= 1
                }
                else if(nums[i]+nums[j]+nums[k]<0){
                    j += 1
                }
                else{
                    a.add(listOf(nums[i], nums[j], nums[k]))
                    j += 1
                    k -= 1
                    while(j<k && nums[j]==nums[j-1]){
                        j+=1
                    }
                    while(k<j && nums[k]==nums[k+1]){
                        k-=1
                    }
                }
            }
        }
        return a
    }
}
